lib/internal/modules/esm/module_job.js

```diff
@@ -121,6 +98,49 @@ class ModuleJob extends ModuleJobBase {
     this.instantiated = undefined;
   }
 
+  /**
+   * Iterates the module requests and links with the loader.
+   * @returns {Promise<ModuleJob[]>} Dependency module jobs.
+   */
+  async _link() {
+    this.module = await this.modulePromise;
+    assert(this.module instanceof ModuleWrap);
+
+    const moduleRequests = this.module.getModuleRequests();
+    // Explicitly keeping track of dependency jobs is needed in order
+    // to flatten out the dependency graph below in `_instantiate()`,
+    // so that circular dependencies can't cause a deadlock by two of
+    // these `link` callbacks depending on each other.
+    // Create an ArrayLike to avoid calling into userspace with `.then`
+    // when returned from the async function.
+    const dependencyJobs = Array(moduleRequests.length);
+    ObjectSetPrototypeOf(dependencyJobs, null);
+
+    // Specifiers should be aligned with the moduleRequests array in order.
+    const specifiers = Array(moduleRequests.length);
```

Comment:
This leads to holey arrays, it might still be better to just use ArrayPrototyprPush since this is in a hot path (I think ArrayPrototypePush can be optimized properly the last time I checked).

Reply:
Do you have a figure for it? As far as I can tell with the latest V8 version, comparing allocating a 20-length array with the two patterns, the result is outstanding that pre-allocating could be ~40% faster: https://jsbench.me/9tlvebajbe/1.

Reply:
The problem with holey arrays come more from the access instead of the creation, since V8 needs to defend against things like this:

Array.prototype[1] = 1;
let a = Array(2);
a[1];  // 1

Reply:
(Also, it's not blocking, I am fine landing with holey arrays, since it's probably still not that much work compared to all the other things the resolution is doing..)

Reply:
Thanks for bringing this up. I think in this particular case the iteration with the c++ api (https://github.com/nodejs/node/blob/main/deps/v8/src/api/api.cc#L8144) difference would be trivial. And the array is not accessed frequently so I'd find pre-allocation would be a good solution here.